client-spec/client-specification-v1.md

```diff
@@ -0,0 +1,322 @@
+# Abstract
+
+This document describes the behaviour that an application must implement
+in order to be a CKAN-compatible mod manager.
+
+Adherence to this document ensures that all mod-management applications
+that consume CKAN metadata behave consistently.
+
+This document refers to version 1 of the CKAN metadata specification.
+
+
+
+# Definitions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in  [RFC 2119](http://www.faqs.org/rfcs/rfc2119.html).
+
+
+
+# Metadata fields
+
+CKAN metadata consists of a JSON file
+conforming to the official JSON schema.
+
+The following sections define how the information contained in the metadata files
+should be interpreted and handled by ckan clients.
+
+
+## Required Fields
+
+The fields described in this section are required by the ckan metadata specification
+and must be present for all valid metadata files.
+A ckan client may assume that all these fields are present in all valid metadata files.
+Clients may assume that all metadata contained in the official CKAN repositories is valid;
+however, client-side validation against the JSON schema is recommended.
+
+### identifier
+
+This is the unique identifier for a module.
+The same identifier is shared by multiple versions of the same module.
+
+A client must use this identifier to discriminate different modules
+and must not allow the user to install two modules with the same identifier.
+
+### version
+
+A string defining the version of the module.
+The specification allows for any string to be used as a version.
+Since there is no standard regarding versioning schemes, a client should not assume any.
+
+As a significant amount of metadata in the official repository
+is obtained automatically from github, leading 'v' characters are common:
+as such, version comparison according to [PEP 440](https://www.python.org/dev/peps/pep-0440/#version-scheme) is recommended.
```

Comment:
There seems to be endless opinions on how we *should* handle version comparisons. The metadata spec explicitly mandates that clients *must* use the Debian version comparison rules, and the C# implementation follows these exactly.

I'm hesitant to change how we do version comparisons for two reasons:

1. We index more than 700 mods. Of those, we've only had a fraction where the Debian versioning spec has not worked, and changing the versioning implementation runs a *very real* risk of breaking existing mods.
1. The netkan pre-processor now has additional options for post-processing of versions into standard forms, or bumping the epoch when versions are truly wacky.
1. Changing versioning schemes is *much* harder than everyone seems to think. If we spot a document implementing spec 1.0–1.6, then we should use the old versioning scheme. If we spot a document with the "new" spec, we should use the new versioning scheme. If we're trying to compare *between* routines, we're screwed.

TL;DR: We're not going to change versioning schemes unless someone can provide *very* compelling reasons, a proof that we wont break existing metadata, and a sane migration plan.

TL;DR;TL;DR: We use the Debian spec.